# Lab 8: Define and Solve an ML Problem of Your Choosing

In [2]:
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns

In this lab assignment, you will follow the machine learning life cycle and implement a model to solve a machine learning problem of your choosing. You will select a data set and choose a predictive problem that the data set supports.  You will then inspect the data with your problem in mind and begin to formulate a  project plan. You will then implement the machine learning project plan. 

You will complete the following tasks:

1. Build Your DataFrame
2. Define Your ML Problem
3. Perform exploratory data analysis to understand your data.
4. Define Your Project Plan
5. Implement Your Project Plan:
    * Prepare your data for your model.
    * Fit your model to the training data and evaluate your model.
    * Improve your model's performance.

## Part 1: Build Your DataFrame

You will have the option to choose one of four data sets that you have worked with in this program:

* The "census" data set that contains Census information from 1994: `censusData.csv`
* Airbnb NYC "listings" data set: `airbnbListingsData.csv`
* World Happiness Report (WHR) data set: `WHR2018Chapter2OnlineData.csv`
* Book Review data set: `bookReviewsData.csv`

Note that these are variations of the data sets that you have worked with in this program. For example, some do not include some of the preprocessing necessary for specific models. 

#### Load a Data Set and Save it as a Pandas DataFrame

The code cell below contains filenames (path + filename) for each of the four data sets available to you.

<b>Task:</b> In the code cell below, use the same method you have been using to load the data using `pd.read_csv()` and save it to DataFrame `df`. 

You can load each file as a new DataFrame to inspect the data before choosing your data set.

In [3]:
# File names of the four data sets
adultDataSet_filename = os.path.join(os.getcwd(), "data", "censusData.csv")
airbnbDataSet_filename = os.path.join(os.getcwd(), "data", "airbnbListingsData.csv")
WHRDataSet_filename = os.path.join(os.getcwd(), "data", "WHR2018Chapter2OnlineData.csv")
bookReviewDataSet_filename = os.path.join(os.getcwd(), "data", "bookReviewsData.csv")

# I am choosing the bookReviewsData.csv dataset for my project.

df = pd.read_csv(bookReviewDataSet_filename )
df.head(10)

Review  Positive Review
0  This was perhaps the best of Johannes Steinhof...             True
1  This very fascinating book is a story written ...             True
2  The four tales in this collection are beautifu...             True
3  The book contained more profanity than I expec...            False
4  We have now entered a second time of deep conc...             True
5  I don't know why it won the National Book Awar...            False
6  The daughter of a prominent Boston doctor is d...            False
7  I was very disapointed in the book.Basicly the...            False
8  I think in retrospect I wasted my time on this...            False
9  I have a hard time understanding what it is th...            False

## Part 2: Define Your ML Problem

Next you will formulate your ML Problem. In the markdown cell below, answer the following questions:

1. List the data set you have chosen.
2. What will you be predicting? What is the label?
3. Is this a supervised or unsupervised learning problem? Is this a clustering, classification or regression problem? Is it a binary classificaiton or multi-class classifiction problem?
4. What are your features? (note: this list may change after your explore your data)
5. Explain why this is an important problem. In other words, how would a company create value with a model that predicts this label?

1. The data set I have chosen is "bookReviewsData.csv"
2. I will be predicting whether a book review is positive or not. My label is 'Positive Review', which will be either true or false, and my only feature for this problem is the 'Review' column.
3. This is a supervised learning problem because for each data point, the 'Review' feature is associated with a 'Positive Review' label. In other words, during the training phase, the ML model is supplied a label along with its features rather than having to draw patterns and groupings by itself. Furthermore, this is a binary classification problem because the label has only one of two outcomes or classes, true or false.
4. As mentioned previously, my feature is "Review". This is the only feature column for my ML problem.
5. This model could be especially useful for websites like Good Reads or publishing companies. Good Reads could used this sentiment analysis model to classify a review left under a book as positive or negative. Furthermore, publishing houses like disney hyperion or wharton could use this model to analyze reviews left for their authors' books and classify them into positive or negative reviews. This model would be a streamlined method for a publishing house to decide the public opinion of their authors' books rather than having employees manually reading through thousands of reviews. Knowing how many positive and negative reviews a given book has can inform the type of books the publishing house publishes in the future, hence can indirectly help boost their profits.

## Part 3: Understand Your Data

The next step is to perform exploratory data analysis. Inspect and analyze your data set with your machine learning problem in mind. Consider the following as you inspect your data:

1. What data preparation techniques would you like to use? These data preparation techniques may include:

    * addressing missingness, such as replacing missing values with means
    * finding and replacing outliers
    * renaming features and labels
    * finding and replacing outliers
    * performing feature engineering techniques such as one-hot encoding on categorical features
    * selecting appropriate features and removing irrelevant features
    * performing specific data cleaning and preprocessing techniques for an NLP problem
    * addressing class imbalance in your data sample to promote fair AI
    

2. What machine learning model (or models) you would like to use that is suitable for your predictive problem and data?
    * Are there other data preparation techniques that you will need to apply to build a balanced modeling data set for your problem and model? For example, will you need to scale your data?
 
 
3. How will you evaluate and improve the model's performance?
    * Are there specific evaluation metrics and methods that are appropriate for your model?
    

Think of the different techniques you have used to inspect and analyze your data in this course. These include using Pandas to apply data filters, using the Pandas `describe()` method to get insight into key statistics for each column, using the Pandas `dtypes` property to inspect the data type of each column, and using Matplotlib and Seaborn to detect outliers and visualize relationships between features and labels. If you are working on a classification problem, use techniques you have learned to determine if there is class imbalance.

<b>Task</b>: Use the techniques you have learned in this course to inspect and analyze your data. You can import additional packages that you have used in this course that you will need to perform this task.

<b>Note</b>: You can add code cells if needed by going to the <b>Insert</b> menu and clicking on <b>Insert Cell Below</b> in the drop-drown menu.

In [4]:
# Routine Data Frame inspections:

print(df.head(10))
print(df.shape)
print(df.columns)

                                              Review  Positive Review
0  This was perhaps the best of Johannes Steinhof...             True
1  This very fascinating book is a story written ...             True
2  The four tales in this collection are beautifu...             True
3  The book contained more profanity than I expec...            False
4  We have now entered a second time of deep conc...             True
5  I don't know why it won the National Book Awar...            False
6  The daughter of a prominent Boston doctor is d...            False
7  I was very disapointed in the book.Basicly the...            False
8  I think in retrospect I wasted my time on this...            False
9  I have a hard time understanding what it is th...            False
(1973, 2)
Index(['Review', 'Positive Review'], dtype='object')


In [5]:
# I am checking if any values in each column are NaN or null
df.isnull().sum()

Review             0
Positive Review    0
dtype: int64

Part 3: Understanding my data:

1. Admittedly, in the data understanding stage, since I don't have any numerical data, there are no trends and patterns I have to infer and visualize. Since this is an NLP problem and I have only one feature column, most of the data preperation techniques listed above like df.describe() and dtypes are not applicable to my problem becuase they pertain to categorical or numerical data. Instead, I will do data cleaning tasks like removing whitespaces, lowercasing all words, removing stopwords and perform pre-processing steps like tokenization or lemmatization before I vectorize the text using TF-IDF.
2. I will be using a feedforward neural network for my problem and using to predict the probability of the Positive Review label being true. Hence, this is a supervised binary classification problem as specified previously. I will vectorize my text using TF-IDF Vectorization. Beyond this, I am not sure I will need to scale or further process my data before modelling.
3. While fitting the model, I will supply parameters such that validation and training loss and accuracy will be displayed. I will use loss and accuracy metrics to evaluate each model configuration to choose the best performing models. I will also evaluate the model's performance on the test set using training loss and accuracy, via the evaluate() method. To keep it simple, I will test 3 different model configurations and choose the best performing model to make final predictions.

## Part 4: Define Your Project Plan

Now that you understand your data, in the markdown cell below, define your plan to implement the remaining phases of the machine learning life cycle (data preparation, modeling, evaluation) to solve your ML problem. Answer the following questions:

* Do you have a new feature list? If so, what are the features that you chose to keep and remove after inspecting the data? 
* Explain different data preparation techniques that you will use to prepare your data for modeling.
* What is your model (or models)?
* Describe your plan to train your model, analyze its performance and then improve the model. That is, describe your model building, validation and selection plan to produce a model that generalizes well to new data. 

1. I do not have a new feature list. The only feature remains 'Review'.
2. I will remove stop words, lowercase all words, and remove whitespaces. Then, I will perform tokenization and perform TF-IDF vectorization of my text. I plan to utilize the parameters already provided in the TF_IDF object to do the pre-processing and data preparation to make the process easier and more streamlined.
3. My model is a feedforward neural network, the simplest neural network, suitable for small data sets. I have chosen this model for ease of implementation and interpretability. A larger corpus and more examples would have required a more complex model.
4. I will vectorize my text data. I will then create an input layer, 3 hidden layers, and an output layer for each model configuration. I will define my loss function and optimizer and use those to do a compile of the model. To analyze each model, I will use training loss and accuracy as well as test loss and accuracy. I will also plot the model's performance over time. I am unsure if GridSearchCV is compatible with neural networks so I will manually test a few hyperparameter configurations. I will change the num epochs and add dropout regularization between hidden layers to create different model configurations. I will select the model with the highest accuracy and lowest loss. Finally, I will make predictions using this final optimal model.

## Part 5: Implement Your Project Plan

<b>Task:</b> In the code cell below, import additional packages that you have used in this course that you will need to implement your project plan.

In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
import tensorflow.keras as keras
import time

<b>Task:</b> Use the rest of this notebook to carry out your project plan. 

You will:

1. Prepare your data for your model.
2. Fit your model to the training data and evaluate your model.
3. Improve your model's performance by performing model selection and/or feature selection techniques to find best model for your problem.

Add code cells below and populate the notebook with commentary, code, analyses, results, and figures as you see fit. 

In [8]:
TfidfVectorizer?

Init signature:
TfidfVectorizer(
    *,
    input='content',
    encoding='utf-8',
    decode_error='strict',
    strip_accents=None,
    lowercase=True,
    preprocessor=None,
    tokenizer=None,
    analyzer='word',
    stop_words=None,
    token_pattern='(?u)\\b\\w\\w+\\b',
    ngram_range=(1, 1),
    max_df=1.0,
    min_df=1,
    max_features=None,
    vocabulary=None,
    binary=False,
    dtype=<class 'numpy.float64'>,
    norm='l2',
    use_idf=True,
    smooth_idf=True,
    sublinear_tf=False,
)
Docstring:     
Convert a collection of raw documents to a matrix of TF-IDF features.

Equivalent to :class:`CountVectorizer` followed by
:class:`TfidfTransformer`.

For an example of usage, see
:ref:`sphx_glr_auto_examples_text_plot_document_classification_20newsgroups.py`.

For an efficiency comparison of the different feature extractors, see
:ref:`sphx_glr_auto_examples_text_plot_hashing_vs_dict_vectorizer.py`.

Read more in the :ref:`User Guide <text_feature_extraction>`.

Parameter

In [9]:
# Creating labelled examples and train, test splits

y = df['Positive Review'] 
X = df['Review']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1234)

# Data Preparation and TF-IDF Vectorization of text data

tfidf_vectorizer = TfidfVectorizer(lowercase=True,stop_words='english')#Tokenization is done by whitespace automatically.
tfidf_vectorizer.fit(X_train)
X_train_tfidf = tfidf_vectorizer.transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

# Creating a vocabulary size

vocabulary_size = len(tfidf_vectorizer.vocabulary_)

In [10]:
# Creating the Neural Network Configuration 1

nn_model = keras.Sequential()

input_layer = keras.layers.InputLayer(input_shape=vocabulary_size)
nn_model.add(input_layer)

hidden_1 = keras.layers.Dense(units=64,activation='relu')
nn_model.add(hidden_1)

hidden_2 = keras.layers.Dense(units=32,activation='relu')
nn_model.add(hidden_2)

hidden_3 = keras.layers.Dense(units=16,activation='relu')
nn_model.add(hidden_3)

output_layer = keras.layers.Dense(units=1,activation='sigmoid')
nn_model.add(output_layer)

nn_model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 64)                1168768   
_________________________________________________________________
dense_1 (Dense)              (None, 32)                2080      
_________________________________________________________________
dense_2 (Dense)              (None, 16)                528       
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 17        
Total params: 1,171,393
Trainable params: 1,171,393
Non-trainable params: 0
_________________________________________________________________


2025-07-30 03:38:54.221384: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2025-07-30 03:38:54.221413: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)
2025-07-30 03:38:54.221487: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (i-015ce9d825cc1dcbc): /proc/driver/nvidia/version does not exist
2025-07-30 03:38:54.221728: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [11]:
# Defining Optimization function, Loss function, and compiling model configuration 1

sgd_optimizer = keras.optimizers.SGD(learning_rate=0.1)
loss_fn = keras.losses.BinaryCrossentropy(from_logits=False)
nn_model.compile(optimizer=sgd_optimizer,loss=loss_fn,metrics=['accuracy'])

In [12]:
# Custom class from Unit 8 assignment to display loss and accuracy and other custom data while the model is running.
class ProgBarLoggerNEpochs(keras.callbacks.Callback):
    
    def __init__(self, num_epochs: int, every_n: int = 50):
        self.num_epochs = num_epochs
        self.every_n = every_n
    
    def on_epoch_end(self, epoch, logs=None):
        if (epoch + 1) % self.every_n == 0:
            s = 'Epoch [{}/ {}]'.format(epoch + 1, self.num_epochs)
            logs_s = ['{}: {:.4f}'.format(k.capitalize(), v)
                      for k, v in logs.items()]
            s_list = [s] + logs_s
            print(', '.join(s_list))

In [13]:
# Fitting Model Configuration 1

t0 = time.time() # start time

num_epochs = 55 # epochs

history = nn_model.fit(X_train_tfidf.toarray(),y_train,epochs=num_epochs,verbose=0,validation_split=0.2,callbacks=[ProgBarLoggerNEpochs(num_epochs, every_n=5)])

t1 = time.time() # stop time

print('Elapsed time: %.2fs' % (t1-t0))

2025-07-30 03:42:00.788627: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:176] None of the MLIR Optimization Passes are enabled (registered 2)
2025-07-30 03:42:00.792655: I tensorflow/core/platform/profile_utils/cpu_utils.cc:114] CPU Frequency: 2649995000 Hz


Epoch [5/ 55], Loss: 0.6917, Accuracy: 0.5114, Val_loss: 0.6919, Val_accuracy: 0.5101
Epoch [10/ 55], Loss: 0.6831, Accuracy: 0.5985, Val_loss: 0.6858, Val_accuracy: 0.7297
Epoch [15/ 55], Loss: 0.6218, Accuracy: 0.7523, Val_loss: 0.6429, Val_accuracy: 0.7230
Epoch [20/ 55], Loss: 0.5068, Accuracy: 0.7667, Val_loss: 0.5430, Val_accuracy: 0.7703
Epoch [25/ 55], Loss: 0.5949, Accuracy: 0.6839, Val_loss: 0.6648, Val_accuracy: 0.5912
Epoch [30/ 55], Loss: 0.3992, Accuracy: 0.8301, Val_loss: 0.4470, Val_accuracy: 0.7838
Epoch [35/ 55], Loss: 0.0301, Accuracy: 1.0000, Val_loss: 0.4584, Val_accuracy: 0.7905
Epoch [40/ 55], Loss: 0.0076, Accuracy: 1.0000, Val_loss: 0.5022, Val_accuracy: 0.7905
Epoch [45/ 55], Loss: 0.0037, Accuracy: 1.0000, Val_loss: 0.5377, Val_accuracy: 0.7770
Epoch [50/ 55], Loss: 0.0023, Accuracy: 1.0000, Val_loss: 0.5713, Val_accuracy: 0.7804
Epoch [55/ 55], Loss: 0.0016, Accuracy: 1.0000, Val_loss: 0.5752, Val_accuracy: 0.7770
Elapsed time: 5.63s


In [15]:
# Evaluating Model Configuration 1's performance on test data

loss, accuracy = nn_model.evaluate(X_test_tfidf.toarray(),y_test)
print('Loss: ', str(loss) , 'Accuracy: ', str(accuracy))

16/16 [==============================] - 0s 1ms/step - loss: 0.5514 - accuracy: 0.8057
Loss:  0.5513968467712402 Accuracy:  0.8056679964065552


In [16]:
# Creating the Neural Network Configuration 2: Dropout Regularization after every hidden layer

nn2_model = keras.Sequential()

input_layer = keras.layers.InputLayer(input_shape=vocabulary_size)
nn2_model.add(input_layer)

hidden_1 = keras.layers.Dense(units=64,activation='relu')
nn2_model.add(hidden_1)
nn2_model.add(keras.layers.Dropout(.25))

hidden_2 = keras.layers.Dense(units=32,activation='relu')
nn2_model.add(hidden_2)
nn2_model.add(keras.layers.Dropout(.25))

hidden_3 = keras.layers.Dense(units=16,activation='relu')
nn2_model.add(hidden_3)
nn2_model.add(keras.layers.Dropout(.25))

output_layer = keras.layers.Dense(units=1,activation='sigmoid')
nn2_model.add(output_layer)

nn2_model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 64)                1168768   
_________________________________________________________________
dropout (Dropout)            (None, 64)                0         
_________________________________________________________________
dense_5 (Dense)              (None, 32)                2080      
_________________________________________________________________
dropout_1 (Dropout)          (None, 32)                0         
_________________________________________________________________
dense_6 (Dense)              (None, 16)                528       
_________________________________________________________________
dropout_2 (Dropout)          (None, 16)                0         
_________________________________________________________________
dense_7 (Dense)              (None, 1)                

In [18]:
# Compiling and fitting model configuration 2

nn2_model.compile(optimizer=sgd_optimizer,loss=loss_fn,metrics=['accuracy'])
t0 = time.time() # start time
num_epochs = 55 # epochs
history = nn2_model.fit(X_train_tfidf.toarray(),y_train,epochs=num_epochs,verbose=0,validation_split=0.2,callbacks=[ProgBarLoggerNEpochs(num_epochs, every_n=5)])
t1 = time.time() # stop time
print('Elapsed time: %.2fs' % (t1-t0))

Epoch [5/ 55], Loss: 0.6924, Accuracy: 0.5080, Val_loss: 0.6924, Val_accuracy: 0.5101
Epoch [10/ 55], Loss: 0.6876, Accuracy: 0.5630, Val_loss: 0.6895, Val_accuracy: 0.5101
Epoch [15/ 55], Loss: 0.6540, Accuracy: 0.6847, Val_loss: 0.6668, Val_accuracy: 0.7669
Epoch [20/ 55], Loss: 0.5460, Accuracy: 0.7473, Val_loss: 0.7031, Val_accuracy: 0.5270
Epoch [25/ 55], Loss: 0.4426, Accuracy: 0.8132, Val_loss: 0.4739, Val_accuracy: 0.7973
Epoch [30/ 55], Loss: 0.3285, Accuracy: 0.8783, Val_loss: 0.4601, Val_accuracy: 0.7838
Epoch [35/ 55], Loss: 0.2550, Accuracy: 0.9341, Val_loss: 0.4528, Val_accuracy: 0.7770
Epoch [40/ 55], Loss: 0.0658, Accuracy: 0.9941, Val_loss: 0.4963, Val_accuracy: 0.7939
Epoch [45/ 55], Loss: 0.0292, Accuracy: 0.9949, Val_loss: 0.5975, Val_accuracy: 0.7804
Epoch [50/ 55], Loss: 0.0155, Accuracy: 0.9983, Val_loss: 0.6912, Val_accuracy: 0.7905
Epoch [55/ 55], Loss: 0.4348, Accuracy: 0.8284, Val_loss: 0.5935, Val_accuracy: 0.6216
Elapsed time: 5.86s


In [19]:
# Evaluating Model Configuration 2's performance on test data

loss, accuracy = nn2_model.evaluate(X_test_tfidf.toarray(),y_test)
print('Loss: ', str(loss) , 'Accuracy: ', str(accuracy))

16/16 [==============================] - 0s 1ms/step - loss: 0.5565 - accuracy: 0.6640
Loss:  0.5565044283866882 Accuracy:  0.6639676094055176


In [20]:
# Creating the Neural Network Configuration 3: Changing the num epochs to 45.

nn3_model = keras.Sequential()

input_layer = keras.layers.InputLayer(input_shape=vocabulary_size)
nn3_model.add(input_layer)

hidden_1 = keras.layers.Dense(units=64,activation='relu')
nn3_model.add(hidden_1)

hidden_2 = keras.layers.Dense(units=32,activation='relu')
nn3_model.add(hidden_2)

hidden_3 = keras.layers.Dense(units=16,activation='relu')
nn3_model.add(hidden_3)

output_layer = keras.layers.Dense(units=1,activation='sigmoid')
nn3_model.add(output_layer)

nn3_model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_8 (Dense)              (None, 64)                1168768   
_________________________________________________________________
dense_9 (Dense)              (None, 32)                2080      
_________________________________________________________________
dense_10 (Dense)             (None, 16)                528       
_________________________________________________________________
dense_11 (Dense)             (None, 1)                 17        
Total params: 1,171,393
Trainable params: 1,171,393
Non-trainable params: 0
_________________________________________________________________


In [21]:
# Compiling and fitting model configuration 3

nn3_model.compile(optimizer=sgd_optimizer,loss=loss_fn,metrics=['accuracy'])
t0 = time.time() # start time
num_epochs = 45 # epochs
history = nn3_model.fit(X_train_tfidf.toarray(),y_train,epochs=num_epochs,verbose=0,validation_split=0.2,callbacks=[ProgBarLoggerNEpochs(num_epochs, every_n=5)])
t1 = time.time() # stop time
print('Elapsed time: %.2fs' % (t1-t0))

Epoch [5/ 45], Loss: 0.6916, Accuracy: 0.5368, Val_loss: 0.6918, Val_accuracy: 0.5101
Epoch [10/ 45], Loss: 0.6761, Accuracy: 0.6517, Val_loss: 0.6826, Val_accuracy: 0.6385
Epoch [15/ 45], Loss: 0.5628, Accuracy: 0.7642, Val_loss: 0.6455, Val_accuracy: 0.5743
Epoch [20/ 45], Loss: 0.4752, Accuracy: 0.7785, Val_loss: 0.8689, Val_accuracy: 0.5236
Epoch [25/ 45], Loss: 0.3114, Accuracy: 0.8639, Val_loss: 0.4451, Val_accuracy: 0.7872
Epoch [30/ 45], Loss: 0.0555, Accuracy: 0.9966, Val_loss: 0.4547, Val_accuracy: 0.7973
Epoch [35/ 45], Loss: 0.0081, Accuracy: 1.0000, Val_loss: 0.5065, Val_accuracy: 0.7973
Epoch [40/ 45], Loss: 0.0037, Accuracy: 1.0000, Val_loss: 0.5255, Val_accuracy: 0.7939
Epoch [45/ 45], Loss: 0.0023, Accuracy: 1.0000, Val_loss: 0.5512, Val_accuracy: 0.7939
Elapsed time: 4.61s


In [22]:
# Evaluating Model Configuration 3's performance on test data

loss, accuracy = nn3_model.evaluate(X_test_tfidf.toarray(),y_test)
print('Loss: ', str(loss) , 'Accuracy: ', str(accuracy))

16/16 [==============================] - 0s 1ms/step - loss: 0.5366 - accuracy: 0.7996
Loss:  0.5365519523620605 Accuracy:  0.7995951175689697


# Model Evaluation and Selection

Out of the 3 model configurations I trained, Model 3 has the least elapsed time at 4.61 s. Model 3 also has the least test loss at approximately 0.54. However, model 1 has the least training loss. Model 1 has the highest test accuracy and both models 1 and 3 reach a training accuracy of 1.000 eventually. I would choose between 1 and 3.